--- echo.c	2018-04-07 15:26:23.558230483 +0530
+++ ../../project_2/project_2.sdk/lwip_test/src/echo.c	2018-04-07 15:20:40.042218250 +0530
@@ -37,6 +37,8 @@
 #include "lwip/tcp.h"
 #if defined (__arm__) || defined (__aarch64__)
 #include "xil_printf.h"
+#include "xil_types.h"
+#include "xil_cache_l.h"
 #endif
 
 int transfer_data() {
@@ -62,6 +64,27 @@
 	/* indicate that the packet has been received */
 	tcp_recved(tpcb, p->len);
 
+	u8 *data = (u8 *)(p->payload);
+	u8 id = data[0];
+	if (data[1] != 0xA5) goto echo; // MAGIC WORD check for data validity
+	u16 len = (data[3] << 8) + data[2];
+	u32 *addr = (data[7] << 24) + (data[6] << 16) + (data[5] << 8) + data[4];
+	if (id == 0) 
+		memcpy(addr,&data[8],len);	
+	else if (id == 1) 
+		memcpy(data,addr,len);
+	else if (id == 2) // IO Write
+		*addr = *((u32 *)&data[8]);
+	else if (id == 4)
+		Xil_L1DCacheFlushRange((u32)addr,len);
+	else 
+		*((u32 *)data) = *addr;
+
+
+	if (id % 2) p->len = len;
+	else {p->len = 1; *data = 1;}
+
+echo:
 	/* echo back the payload */
 	/* in this case, we assume that the payload is < TCP_SND_BUF */
 	if (tcp_sndbuf(tpcb) > p->len) {
